api/clients/v2/cert_builder.go

```diff
@@ -0,0 +1,174 @@
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/Layr-Labs/eigenda/api/clients/v2/coretypes"
+	coreEth "github.com/Layr-Labs/eigenda/core/eth"
+
+	disperser "github.com/Layr-Labs/eigenda/api/grpc/disperser/v2"
+	"github.com/Layr-Labs/eigenda/common"
+	certTypesBinding "github.com/Layr-Labs/eigenda/contracts/bindings/IEigenDACertTypeBindings"
+	opsrbinding "github.com/Layr-Labs/eigenda/contracts/bindings/OperatorStateRetriever"
+	"github.com/Layr-Labs/eigenda/core"
+	coreV2 "github.com/Layr-Labs/eigenda/core/v2"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	gethcommon "github.com/ethereum/go-ethereum/common"
+)
+
+
+type CertBuilder struct {
+	logger 					logging.Logger
+	opsrCaller				*opsrbinding.ContractOperatorStateRetrieverCaller
+	registryCoordinatorAddr gethcommon.Address
+}
+
+// NewCertBuilder constructs a new CertBuilder instance used to build EigenDA certificates
+// across different versions.
+func NewCertBuilder(
+	logger logging.Logger,
+	opsrAddr gethcommon.Address,
+	registryCoordinatorAddr gethcommon.Address,
+	ethClient common.EthClient,
+) (*CertBuilder, error) {
+	if logger == nil {
+		return nil, fmt.Errorf("logger cannot be nil")
+	}
+	// Create the Operator State Retriever caller
+	opsrCaller, err := opsrbinding.NewContractOperatorStateRetrieverCaller(opsrAddr, ethClient)
+	if err != nil {
+		return nil, fmt.Errorf("create operator state retriever caller: %w", err)
+	}
+	
+
+	return &CertBuilder{
+		logger: logger,
+		opsrCaller: opsrCaller,
+		registryCoordinatorAddr: registryCoordinatorAddr,
+	}, nil
+}
+
+// BuildCert builds an EigenDA certificate of the specified version using the provided blob key and blob status reply.
+func (cb *CertBuilder) BuildCert(
+	ctx context.Context,
+	certVersion coretypes.CertificateVersion,
+	blobKey coreV2.BlobKey,
+	blobStatusReply *disperser.BlobStatusReply,
+) (coretypes.EigenDACert, error) {
+	switch certVersion {
+	case coretypes.VersionThreeCert:
+		return cb.buildEigenDAV3Cert(ctx, blobKey, blobStatusReply)
+	default:
+		return nil, fmt.Errorf("unsupported EigenDA cert version: %d", certVersion)
+	}
+}
+
+// buildEigenDAV3Cert builds an EigenDA certificate of version 3 using the provided blob key and blob status reply.
+func (cb *CertBuilder) buildEigenDAV3Cert(
+	ctx context.Context,
+	blobKey coreV2.BlobKey,
+	blobStatusReply *disperser.BlobStatusReply,
+) (*coretypes.EigenDACertV3, error) {
+	// TODO: configurable timeout
+	timeoutCtx, cancel := context.WithTimeout(ctx, time.Second)
+	defer cancel()
+	nonSignerStakesAndSignature, err := cb.getNonSignerStakesAndSignature(
+		timeoutCtx, blobStatusReply.GetSignedBatch())
+	if err != nil {
+		return nil, fmt.Errorf("get non signer stake and signature: %w", err)
+	}
+	cb.logger.Debug("Retrieved NonSignerStakesAndSignature", "blobKey", blobKey.Hex())
+
+	eigenDACert, err := coretypes.BuildEigenDACertV3(blobStatusReply, nonSignerStakesAndSignature)
```

Comment:
why do we need `buildEigenDAV3Cert` and `BuildEigenDACertV3`? Its late and I'm super tired so prob missing something but seems like a lot of layers. Also shouldnt they be called the same thing (v3 in different places right now)?